src/global/MethodTest.java

```diff
@@ -0,0 +1,241 @@
+package global;
+
+import global.exceptions.InvalidClauseException;
+import global.variables.Clause;
+import global.variables.RandomClause;
+import global.variables.clauses.PlaceHolder;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Set;
+import java.util.regex.Matcher;
+
+import static global.tools.CustomAssertions._assertTrue;
+import static global.tools.CustomAssertions._fail;
+import static global.tools.Logger.parseTestInformation;
+import static global.utils.RegexUtil.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+
+public class MethodTest {
+    private ByteArrayOutputStream methodOutput;
+    private final Class<?> methodClass;
+    private final String methodName;
+    private final Object[] arguments;
+    private final Class<?>[] methodArgumentTypes;
+    private final Clause[] methodTestSentence;
+    private String methodNotFoundErrorMessage;
+    private String incorrectMethodStructureErrorMessage;
+
+    public MethodTest(Class<?> methodClass, String methodName, Object[] arguments, Class<?>[] methodArgumentTypes) {
+        this.methodClass = methodClass;
+        this.methodName = methodName;
+        this.arguments = arguments;
+        this.methodArgumentTypes = methodArgumentTypes;
+        this.methodTestSentence = null;
+        this.incorrectMethodStructureErrorMessage = "Your " + methodName + " method's output did not follow the correct structure/syntax";
+        setMethodOutput();
+    }
+
+    public MethodTest(Class<?> methodClass, String methodName) {
+        this.methodClass = methodClass;
+        this.methodName = methodName;
+        this.arguments = null;
+        this.methodArgumentTypes = null;
+        this.methodTestSentence = null;
+        this.incorrectMethodStructureErrorMessage = "Your " + methodName + " method's output did not follow the correct structure/syntax";
+        setMethodOutput();
+    }
+
+    public MethodTest(Class<?> methodClass, String methodName, Object[] arguments, Class<?>[] methodArgumentTypes, Clause[] methodTestSentence) {
```

Comment:
@keyvankhademi

Reply:
Hmm! This definitely looks better when calling the method. On the other hand it might make the implementation more dirty (maybe not).

@Hedgemon4 Let's go for it.

Reply:
Sounds good, I will work on designing this.